File: nms.cpp
#include <vector>
#include <algorithm>

using std::min;
using std::max;
using std::vector;
using std::remove_if;

struct bbox {
	int x1, x2, y1, y2;
	int label;
	float score;
};

float bbox_iou(const bbox& bbox1, const bbox &bbox2) {
	int xmin = max(bbox1.x1, bbox2.x1);
	int ymin = max(bbox1.y1, bbox2.y2);
	int xmax = min(bbox1.x2, bbox2.y2);
	int ymax = min(bbox1.y2, bbox2.y2);
	float inter_area = max((xmax - xmin), 0) * max((ymax - ymin), 0);
	float area1 = (bbox1.x2 - bbox1.x1) * (bbox1.y2 - bbox1.y1);
	float area2 = (bbox2.x2 - bbox2.x1) * (bbox2.y2 - bbox2.y1);
	auto uniou_area = inter_area / (area1 + area2 + inter_area);
	if (uniou_area == 0) return 0;
	float iou = inter_area / uniou_area;
	return iou;
}

float nms(vector<bbox>& bboxes, const float score_thr, const float iou_thr) {
	bboxes.erase(remove_if(bboxes.begin(), bboxes.end(), [](const bbox &bbox) {
		return bbox.score < 0.05;
	}), bboxes.end());

	sort(bboxes.begin(), bboxes.end(), [](const bbox& bbox1, const bbox& bbox2) {
		return bbox1.score > bbox2.score;
	});
	for (int i=0; i<bboxes.size(); i++) {
		for (int j=i+1; j < bboxes.size(); j++){
			if (bboxes[i].label == bboxes[j].label) {
				auto iou = bbox_iou(bboxes[i], bboxes[j]);
				if (iou > iou_thr) {
					bboxes[j].score = 0;
				}
			}
		}
	}
	bboxes.erase(remove_if(bboxes.begin(), bboxes.end(), [](const bbox& bbox) {
		return bbox.score == 0;
	}), bboxes.end());
}

File: nms.py
import numpy as np

def bbox_iou(bbox1:np.ndarray, bbox2:np.ndarray) -> float:
    xmin = max(bbox1[0], bbox2[0])
    xmax = min(bbox1[2], bbox2[2])
    ymin = max(bbox1[1], bbox2[1])
    ymax = min(bbox1[3], bbox2[3])
    inter_area = (xmax - xmin) * (ymax - ymin)
    area1 = (bbox1[2] - bbox1[0]) * (bbox1[3] - bbox1[1])
    area2 = (bbox2[2] - bbox2[0]) * (bbox2[3] - bbox2[1])
    union = area1 + area2 - inter_area
    return inter_area / (union + 1e-6)

def nms(bboxes:np.ndarray, scores:np.ndarray, score_thr:float, iou_thr:float) -> np.ndarray:
    valid = scores > score_thr
    bboxes = bboxes[valid]
    scores = scores[valid]
    inds = scores.argsort()
    inds = inds[::-1]
    keep_inds = []
    while inds:
        cur = inds[0]
        cur_score = scores[cur]
        cur_bbox = bboxes[cur]
        keep = True 
        for ind in keep_inds:
            bbox = bboxes[ind]
            iou = bbox_iou(bbox, cur_bbox)
            if iou > iou_thr:
                keep = False
                break
            if keep:
                keep_inds.append(cur)
            inds = inds[1:]
    return np.array(keep_inds)